sdk/storage/azblob/internal/shared/challenge_policy_test.go

```diff
@@ -0,0 +1,85 @@
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+	"github.com/Azure/azure-sdk-for-go/sdk/internal/mock"
+	"github.com/stretchr/testify/require"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+)
+
+type credentialFunc func(context.Context, policy.TokenRequestOptions) (azcore.AccessToken, error)
+
+func (cf credentialFunc) GetToken(ctx context.Context, options policy.TokenRequestOptions) (azcore.AccessToken, error) {
+	return cf(ctx, options)
+}
+
+func TestChallengePolicy(t *testing.T) {
+	accessToken := "***"
+	storageResource := "https://storage.azure.com"
+	storageScope := "https://storage.azure.com/.default"
+	challenge := `Bearer authorization_uri="https://login.microsoftonline.com/{tenant}", resource_id="{storageResource}"`
+	diskResource := "https://disk.azure.com/"
+	diskScope := "https://disk.azure.com/.default"
+
+	for _, test := range []struct {
+		expectedScope, format, resource string
+	}{
+		{format: challenge, resource: storageResource, expectedScope: storageScope},
+		{format: challenge, resource: storageResource, expectedScope: storageScope},
+		{format: challenge, resource: diskResource, expectedScope: diskScope},
+		{format: challenge, resource: diskResource, expectedScope: diskScope},
+	} {
+		t.Run("", func(t *testing.T) {
+			srv, close := mock.NewServer(mock.WithTransformAllRequestsToTestServerUrl())
+			defer close()
+			srv.AppendResponse(
+				mock.WithHeader("WWW-Authenticate", strings.ReplaceAll(test.format, "{storageResource}", test.resource)),
+				mock.WithStatusCode(401),
+			)
+			srv.AppendResponse(mock.WithPredicate(func(r *http.Request) bool {
+				if authz := r.Header.Values("Authorization"); len(authz) != 1 || authz[0] != "Bearer "+accessToken {
+					t.Errorf(`unexpected Authorization "%s"`, authz)
+				}
+				return true
+			}))
+			srv.AppendResponse()
+			authenticated := false
+			cred := credentialFunc(func(ctx context.Context, tro policy.TokenRequestOptions) (azcore.AccessToken, error) {
+				authenticated = true
+				require.Equal(t, []string{test.expectedScope}, tro.Scopes)
+				return azcore.AccessToken{Token: accessToken, ExpiresOn: time.Now().Add(time.Hour)}, nil
+			})
+			p := NewStorageChallengePolicy(cred)
+			pl := runtime.NewPipeline("", "",
+				runtime.PipelineOptions{PerRetry: []policy.Policy{p}},
+				&policy.ClientOptions{Transport: srv},
+			)
+			req, err := runtime.NewRequest(context.Background(), "GET", "https://localhost")
+			require.NoError(t, err)
+			_, err = pl.Do(req)
+			require.True(t, authenticated, "policy should have authenticated")
+		})
+	}
+}
+
+func TestParseTenant(t *testing.T) {
+	actual := parseTenant("")
+	require.Empty(t, actual)
+
+	expected := "00000000-0000-0000-0000-000000000000"
+	sampleURL := "https://login.microsoftonline.com/" + expected
+	actual = parseTenant(sampleURL)
+	require.Equal(t, expected, actual, "tenant was not properly parsed, got %s, expected %s", actual, expected)
```

Comment:
When using `require.Equal()`, if the values don't match, the function will print out got/want for you.

Reply:
Gotcha, will modify this message